src/java.desktop/windows/native/libawt/windows/ShellFolder2.cpp

```diff
@@ -980,10 +980,17 @@ JNIEXPORT jlong JNICALL Java_sun_awt_shell_Win32ShellFolder2_extractIcon
         UINT uFlags = getDefaultIcon ? GIL_DEFAULTICON : GIL_FORSHELL | GIL_ASYNC;
         hres = pIcon->GetIconLocation(uFlags, szBuf, MAX_PATH, &index, &flags);
         if (SUCCEEDED(hres)) {
+            HICON hIconSmall;
             if (size < 24) {
                 size = 16;
             }
-            hres = pIcon->Extract(szBuf, index, &hIcon, NULL, size);
+            hres = pIcon->Extract(szBuf, index, &hIcon, &hIconSmall, size);
```

Comment:
Looking into this change I assume that the reason why the windows "ignores" the requested size of 16x16 was that we request the size 16x16 but after that used the large icon only which is 32x32 by default, is it correct?

The last parameter of [this](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-iextracticonw-extract) method:
>nIconSize
The desired size of the icon, in pixels. The low word contains the size of the large icon, and the high word contains the size of the small icon. The size specified can be the width or height. The width of an icon always equals its height.

Do we need to update the size value and take into account the low/hi sizes?